File: android/src/main/java/com/reactnativelecomscan/LecomScanModule.java
package com.reactnativelecomscan;

import androidx.annotation.NonNull;
import android.app.Activity;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.RCTNativeAppEventEmitter;
import android.device.ScanDevice;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;

@ReactModule(name = LecomScanModule.NAME)
public class LecomScanModule extends ReactContextBaseJavaModule {
  ScanDevice sd = new ScanDevice();
  public static final String NAME = "LecomScan";
  private static final String SCAN_ACTION = "scan.rcv.message";
  private Boolean isScanning = false;
  private final ReactApplicationContext mContext;

  public LecomScanModule(ReactApplicationContext reactContext) {
    super(reactContext);
    mContext = reactContext;
  }

  private final BroadcastReceiver mScanReceiver = new BroadcastReceiver() {
    @Override
    public void onReceive(Context context, Intent intent) {
      String action = intent.getAction();
      if (SCAN_ACTION.equals(action)) {
        byte[] barcode = intent.getByteArrayExtra("barocode");
        int barcodeLen = intent.getIntExtra("length", 0);
        String temp = intent.getStringExtra("barcodeType");
        String aimID = intent.getStringExtra("aimid");
        String barcodeStr = new String(barcode, 0, barcodeLen);
        mContext.getJSModule(RCTNativeAppEventEmitter.class)
          .emit("EventLecomScanSuccess", barcodeStr);
        sd.stopScan();
      }
    }
  };

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }

  @ReactMethod
  public void init() {
    final Activity activity = getCurrentActivity();
    sd = new ScanDevice();
    sd.setOutScanMode(0);  // Mode-value: 0 broadcast mode
    IntentFilter filter = new IntentFilter();
    filter.addAction(SCAN_ACTION);
    activity.registerReceiver(mScanReceiver, filter);
  }

  @ReactMethod
  public void toggleScan() {
    final Activity activity = getCurrentActivity();
    if (isScanning) {
      sd.stopScan();
      activity.unregisterReceiver(mScanReceiver);
      isScanning = false;
    } else {
      sd.startScan();
      IntentFilter filter = new IntentFilter();
      filter.addAction(SCAN_ACTION);
      activity.registerReceiver(mScanReceiver, filter);
      isScanning = true;
    }
  }

  // Required for rn built in EventEmitter Calls.
  @ReactMethod
  public void addListener(String eventName) {
  }

  @ReactMethod
  public void removeListeners(Integer count) {
  }
}

File: android/src/main/java/com/lecomscan/LecomScanModule.kt
package com.lecomscan

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.device.ScanDevice
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.modules.core.RCTNativeAppEventEmitter

@ReactModule(name = LecomScanModule.NAME)
class LecomScanModule internal constructor(private val mContext: ReactApplicationContext) :
  LecomScanSpec(mContext), LifecycleEventListener {

  private var sd: ScanDevice? = null
  private val SCAN_ACTION = "scan.rcv.message"
  private var isScanning = false
  private var isReceiverRegistered = false

  init {
    mContext.addLifecycleEventListener(this)
  }

  private val handler = Handler(Looper.getMainLooper())

  // BroadcastReceiver to handle scan results
  private val mScanReceiver: BroadcastReceiver = object : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
      val action = intent.action
      if (SCAN_ACTION == action) {
        val barcode = intent.getByteArrayExtra("barocode")
        val barcodeStr = barcode?.let {
          val barcodeLen = intent.getIntExtra("length", 0)
          String(it, 0, barcodeLen)
        }
        barcodeStr?.let {
          handler.post {
            val emitter = mContext.getJSModule(RCTNativeAppEventEmitter::class.java)
            emitter.emit("EventLecomScanSuccess", barcodeStr)
          }
        }
        sd?.stopScan()
        isScanning = false
      }
    }
  }

  override fun getName(): String {
    return NAME
  }

  // Register BroadcastReceiver for scan results
  private fun registerReceiver() {
    if (!isReceiverRegistered) {
      val filter = IntentFilter()
      filter.addAction(SCAN_ACTION)
      mContext.registerReceiver(mScanReceiver, filter)
      isReceiverRegistered = true
    }
  }

  // Unregister BroadcastReceiver for scan results
  private fun unregisterReceiver() {
    if (isReceiverRegistered) {
      try {
        mContext.unregisterReceiver(mScanReceiver)
      } catch (e: IllegalArgumentException) {
        // Handle the case where the receiver was not registered
      }
      isReceiverRegistered = false
    }
  }

  private fun isScanDeviceAvailable(): Boolean {
    return try {
      Class.forName("android.device.ScanDevice")
      true
    } catch (e: ClassNotFoundException) {
      false
    }
  }

  // Initialize the ScanDevice and register the receiver
  @ReactMethod
  override fun init() {
    try {
      if (isScanDeviceAvailable()) {
        if (sd == null) {
          sd = ScanDevice()
          sd?.setOutScanMode(0)  // Mode-value: 0 for broadcast mode
        }
        registerReceiver()
      } else {
        // Optionally, emit an event or log that the ScanDevice is not available
      }
    } catch (e: Exception) {
      Log.e("LecomScanModule", "Error initializing scanner: $e")
    }
  }

  // Stop the ScanDevice and unregister the receiver
  @ReactMethod
  override fun stop() {
    sd?.stopScan()
    sd = null
    unregisterReceiver()
    isScanning = false
  }

  // Toggle scanning: Start or stop based on current state
  @ReactMethod
  override fun toggleScan() {
    if (isScanning) {
      stop()
    } else {
      if (sd == null) {
        sd = ScanDevice()
        sd?.setOutScanMode(0)
      }
      sd?.startScan()
      isScanning = true
    }
  }

  // Required for rn built-in EventEmitter Calls
  @ReactMethod
  override fun addListener(eventName: String) {
    // No operation needed here, just for consistency with JS EventEmitter
  }

  @ReactMethod
  override fun removeListeners(count: Double) {
    // No operation needed here, just for consistency with JS EventEmitter
  }

  override fun onHostResume() {
    if (isScanning) {
      sd?.startScan()
    }
  }

  override fun onHostPause() {
    if (isScanning) {
      sd?.stopScan()
    }
  }

  override fun onHostDestroy() {
    // Activity has been destroyed
    stop()
  }

  override fun onCatalystInstanceDestroy() {
    super.onCatalystInstanceDestroy()
    stop()
    mContext.removeLifecycleEventListener(this)
  }

  companion object {
    const val NAME = "LecomScan"
  }
}
